secureboot_policy.go

```diff
@@ -696,88 +696,89 @@ type secureBootDbSet struct {
-// are copied and can be mutated in the copy, the individual digests are not copied and should be considered read only. The pointers
-// to secureBootDb structs are copied and can be changed in the copy, but the actual secureBootDb instances should also be considered
-// read only.
-func (p *secureBootPolicyGenPath) duplicate() *secureBootPolicyGenPath {
-	n := &secureBootPolicyGenPath{}
+// branch creates a branch point in the current branch if one doesn't exist already (although inserting this branch point with
+// PCRProtectionProfile.AddProfileOR is deferred until later), and creates a new sub-branch at the current branch point. Once
+// this has been called, no more instructions can be inserted in to the current branch.
+func (p *secureBootPolicyGenBranch) branch() *secureBootPolicyGenBranch {
+	c := &secureBootPolicyGenBranch{
+		profile:       NewPCRProtectionProfile(),
+		dbUpdateLevel: p.dbUpdateLevel,
+		dbSet:         p.dbSet}
 
-	n.pcrValue = make(tpm2.Digest, len(p.pcrValue))
-	copy(n.pcrValue, p.pcrValue)
+	c.firmwareVerificationEvents = make(tpm2.DigestList, len(p.firmwareVerificationEvents))
+	copy(c.firmwareVerificationEvents, p.firmwareVerificationEvents)
+	c.shimVerificationEvents = make(tpm2.DigestList, len(p.shimVerificationEvents))
+	copy(c.shimVerificationEvents, p.shimVerificationEvents)
 
-	n.dbUpdateLevel = p.dbUpdateLevel
-	n.dbSet = p.dbSet
+	p.subBranches = append(p.subBranches, c)
 
-	n.firmwareVerificationEvents = make(tpm2.DigestList, len(p.firmwareVerificationEvents))
-	copy(n.firmwareVerificationEvents, p.firmwareVerificationEvents)
-	n.shimVerificationEvents = make(tpm2.DigestList, len(p.shimVerificationEvents))
-	copy(n.shimVerificationEvents, p.shimVerificationEvents)
+	return c
+}
 
-	return n
+type sbLoadEventAndBranches struct {
+	event    *EFIImageLoadEvent
+	branches []*secureBootPolicyGenBranch
 }
 
-type loadEventAndPaths struct {
-	event *EFIImageLoadEvent
-	paths []*secureBootPolicyGenPath
+func (e *sbLoadEventAndBranches) branch(event *EFIImageLoadEvent) *sbLoadEventAndBranches {
+	var branches []*secureBootPolicyGenBranch
+	for _, b := range e.branches {
+		if b.profile == nil {
+			continue
+		}
+		branches = append(branches, b.branch())
+	}
+	return &sbLoadEventAndBranches{event, branches}
 }
 
 // secureBootPolicyGen is the main structure involved with computing secure boot policy PCR digests.
 type secureBootPolicyGen struct {
 	*EFISecureBootPolicyProfileParams
 }
 
-// extendMeasurement extends the supplied digest to the current value of pcrValue for the specified event path.
-func (g *secureBootPolicyGen) extendMeasurement(path *secureBootPolicyGenPath, digest tpm2.Digest) {
-	path.extendMeasurement(g.PCRAlgorithm, digest)
+// extendMeasurement extends the supplied digest in to the specified branch.
+func (g *secureBootPolicyGen) extendMeasurement(branch *secureBootPolicyGenBranch, digest tpm2.Digest) {
```

Comment:
shouldn't some of this lowest level methods taking only a branch * and not using the params maybe live on secureBootPolicyGenBranch itself now?

Reply:
I've moved all of the methods that take a single branch to secureBootPolicyGenBranch now.